core/services/workflows/engine.go

```diff
@@ -134,204 +165,335 @@ func (e *Engine) registerTrigger(ctx context.Context) error {
-        - "0x1111111111111111111100000000000000000000000000000000000000000000" # ETHUSD
-        - "0x2222222222222222222200000000000000000000000000000000000000000000" # LINKUSD
-        - "0x3333333333333333333300000000000000000000000000000000000000000000" # BTCUSD
-        
-consensus:
-  - type: "offchain_reporting"
-    ref: evm_median
-    inputs:
-      observations:
-        - $(report_data.outputs)
-    config:
-      aggregation_method: data_feeds_2_0
-      aggregation_config:
-        0x1111111111111111111100000000000000000000000000000000000000000000:
-          deviation: "0.001"
-          heartbeat: "30m"
-        0x2222222222222222222200000000000000000000000000000000000000000000:
-          deviation: "0.001"
-          heartbeat: "30m"
-        0x3333333333333333333300000000000000000000000000000000000000000000:
-          deviation: "0.001"
-          heartbeat: "30m"
-      encoder: EVM
-      encoder_config:
-        abi: "mercury_reports bytes[]"
-
-targets:
-  - type: write_polygon-testnet-mumbai
-    inputs:
-      report:
-        - $(evm_median.outputs.reports)
-    config:
-      address: "0x3F3554832c636721F1fD1822Ccca0354576741Ef"
-      params: [($inputs.report)]
-      abi: "receive(report bytes)"
-  - type: write_ethereum-testnet-sepolia
-    inputs:
-      report:
-        - $(evm_median.outputs.reports)
-    config:
-      address: "0x54e220867af6683aE6DcBF535B4f952cB5116510"
-      params: ["$(inputs.report)"]
-      abi: "receive(report bytes)"
-`
-
-	workflow, err := Parse(yamlWorkflowSpec)
+type Config struct {
+	Spec           string
+	Lggr           logger.Logger
+	Registry       types.CapabilitiesRegistry
+	MaxWorkerLimit int
+}
+
+const (
+	defaultWorkerLimit = 100
+)
+
+func NewEngine(cfg Config) (engine *Engine, err error) {
+	if cfg.MaxWorkerLimit == 0 {
+		cfg.MaxWorkerLimit = defaultWorkerLimit
+	}
+	// TODO: validation of the workflow spec
+	// We'll need to check, among other things:
+	// - that there are no node `ref` called `trigger` as this is reserved for any triggers
+	// - that there are no duplicate `ref`s
+	// - that the `ref` for any triggers is empty -- and filled in with `trigger`
+	// - etc.
+
+	workflow, err := Parse(cfg.Spec)
 	if err != nil {
 		return nil, err
 	}
+
+	// Instantiate semaphore to put a limit on the number of workers
+	newWorkerCh := make(chan struct{}, cfg.MaxWorkerLimit)
+	for i := 0; i < cfg.MaxWorkerLimit; i++ {
+		newWorkerCh <- struct{}{}
+	}
+
 	engine = &Engine{
-		logger:     lggr.Named("WorkflowEngine"),
-		registry:   registry,
-		workflow:   workflow,
-		callbackCh: make(chan capabilities.CapabilityResponse),
+		logger:       cfg.Lggr.Named("WorkflowEngine"),
+		registry:     cfg.Registry,
+		workflow:     workflow,
+		store:        newStore(),
+		queue:        newQueue[stepRequest](),
+		newWorkerCh:  newWorkerCh,
+		stepUpdateCh: make(chan stepState),
+		callbackCh:   make(chan capabilities.CapabilityResponse),
+		stopCh:       make(chan struct{}),
```

Comment:
Some questions in the spirit of seeking simplicity ;)
1. Do we have a good reason to use a thread pool and if not can we get rid of it?
2. Do we have a good reason to implement a custom queue and if not can we get rid of it?
3. Do we have a good reason to have 3 different ways for event notifications (callbackCh, queue.out, stepUpdateCh) and if not can we reduce it to 2 or 1?

Reply:
1. Yes, to avoid runaway resource consumption. If we don't care about that, I can remove.
2. The queue is probably not needed, I added it since I thought it could be helpful to peek from the queue in some situations, which channels won't allow you to do. However, we don't actually use this functionality atm, so can easily add it later.
3. They're different messages and are used for different purposes (callbackCh to receive trigger events, queue.out to receive stepRequests, stepUpdateCh to process step updates). I think we definitely need callbackCh to be separate, we can arguably combine queue.out and stepUpdateCh with an overarching message type and switch underneath on the message type. Which of those we pick feels like a matter of preference to me tbh. In go channels are so cheap that I don't see why we'd want to reduce the number unless we somehow manage to reduce the number message types (which I don't think is possible)